tslearn/neighbors.py

```diff
@@ -12,16 +12,75 @@
 from sklearn.utils.validation import check_is_fitted
 from scipy.spatial.distance import cdist as scipy_cdist
 
-from tslearn.metrics import cdist_dtw, cdist_soft_dtw, VARIABLE_LENGTH_METRICS
+from tslearn.metrics import cdist_dtw, cdist_soft_dtw, \
+    cdist_sax, VARIABLE_LENGTH_METRICS
+from tslearn.piecewise import SymbolicAggregateApproximation
 from tslearn.utils import (to_time_series_dataset, to_sklearn_dataset,
                            check_dims)
 
-neighbors.VALID_METRICS['brute'].extend(['dtw', 'softdtw'])
+neighbors.VALID_METRICS['brute'].extend(['dtw', 'softdtw', 'sax'])
 
 
 class KNeighborsTimeSeriesMixin(KNeighborsMixin):
     """Mixin for k-neighbors searches on Time Series."""
 
+    def _sax_preprocess(self, X, n_segments=None, alphabet_size_avg=None):
+        # First, we standard-normalize the data using mu and sigma
+        # of entire dataset, as opposed to timeseries-wise.
+        if not hasattr(self, '_sax_mu') or self._sax_mu is None:
+            self._sax_mu = numpy.mean(X)
+        if not hasattr(self, '_sax_sigma') or self._sax_sigma is None:
+            self._sax_sigma = numpy.std(X)
+
+        X = (X - self._sax_mu) / self._sax_sigma
+
+        # Now SAX-transform the timeseries
+        if not hasattr(self, '_sax') or self._sax is None:
+            self._sax = SymbolicAggregateApproximation(
+                n_segments=n_segments,
+                alphabet_size_avg=alphabet_size_avg
+            )
+
+        X = to_time_series_dataset(X)
+        X = self._sax.fit_transform(X)
+
+        return X
+
+    def _calculate_X_(self, X, other_X=None):
```

Comment:
could maybe use a more explicit name, such as `_precompute_cross_distances`?

Reply:
That is indeed a much better and clear name!